src/main.rs

```diff
@@ -937,21 +938,69 @@ fn do_cmd_certify(
                 });
         }
         CertifyKind::Delta { from, to } => {
-            let importable = from.git_rev.is_none() && to.git_rev.is_none();
+            let from_is_git_version = from.git_rev.is_some();
+            let kind = AuditKind::Delta { from, to };
+            let importable = kind.default_importable();
+
+            let mut entry = AuditEntry {
+                kind,
+                criteria,
+                who,
+                importable,
+                notes,
+                aggregated_from: vec![],
+                is_fresh_import: false,
+            };
+
+            // Collapse a delta audit with a git `from` version with a prior audit that is
+            // non-importable and has identical and satisfied criteria.
+            //
+            // We merge an adjacent audit for a prior version with the new audit (updating the new
+            // audit). The later `update_store` call will remove the now-unused prior audit.
+            if from_is_git_version && !sub_args.no_collapse {
+                // A closure which returns whether the given audit entry satisfies the criteria
+                // being certified.
+                let is_rooted_for_criteria = {
+                    let mapper = CriteriaMapper::new(&store.audits.criteria);
+                    let criteria = mapper.criteria_from_list(&entry.criteria);
+                    let audit_graph = resolver::AuditGraph::build(store, &mapper, &package, None)
+                        .map_err(|_| CertifyError::BadAuditGraph)?;
```

Comment:
I think it might be nice to fall over gracefully in this scenario, and instead not collapse if we fail to build an audit graph. 

IIRC generally this failure is due to a violation entry, and giving a generic error which doesn't describe the violation issue like this seems like it'd be a poor user experience.

Reply:
Should we display a warning when that happens? I'm just concerned someone _expecting_ a collapse that doesn't occur would be confused.

Reply:
At some point I'd like to factor detecting violation errors out of audit graph building (as it's effectively an unrelated issue) so we'd be able to build an audit graph infallibly (we've actually successfully built one and we just discard it when we return errors)

https://github.com/mozilla/cargo-vet/blob/5cfaa7fe2d1eec82ad9fc5058894f655f55c2c84/src/resolver.rs#L1360-L1368

I suppose we could log a warning for now when it happens though, or we could do the refactor now and make the `build` function return a pair like `(Self, Vec<ViolationConflict>)`